cpp/arcticdb/processing/aggregation.cpp

```diff
@@ -489,4 +509,157 @@ SegmentInMemory CountAggregatorData::finalize(const ColumnName& output_column_na
+            auto col_data = input_column->column_->data();
+            auto out_ptr = reinterpret_cast<GlobalRawType*>(aggregated_.data());
+            entity::details::visit_type(input_column->column_->type().data_type(), [this, &groups, &out_ptr, &col_data] (auto type_desc_tag) {
+                using ColumnTagType = std::decay_t<decltype(type_desc_tag)>;
+                using ColumnType =  typename ColumnTagType::raw_type;
+                auto groups_pos = 0;
+                auto grp_idx = groups[0];
+                while (auto block = col_data.next<TypeDescriptorTag<ColumnTagType, DimensionTag<entity::Dimension::Dim0>>>()) {
+                    auto ptr = reinterpret_cast<const ColumnType *>(block.value().data());
+                    for (auto i = 0u; i < block.value().row_count(); ++i, ++ptr) {
+                        auto& val = out_ptr[groups[groups_pos]];
+                        bool is_first_group_el = (groups_cache_.find(groups[groups_pos]) == groups_cache_.end());
+                        if constexpr(std::is_floating_point_v<ColumnType>) {
+                            const auto& curr = GlobalRawType(*ptr);
+                            if ((grp_idx == groups[groups_pos]) && (is_first_group_el || std::isnan(static_cast<ColumnType>(val)))) {
+                                groups_cache_.insert(groups[groups_pos]);
+                                val = curr;
+                            }
+                            else if (grp_idx != groups[groups_pos]) { // Updating when changing group
+                                if (is_first_group_el) { // Group not seen yet
+                                    groups_cache_.insert(groups[groups_pos]);
+                                    val = curr;
+                                }
+                                else {
+                                    if (std::isnan(static_cast<ColumnType>(val))) {
+                                        val = curr;
+                                    }
+                                }
+                                grp_idx = groups[groups_pos];
+                            }
+                        } else {
+                            const auto& curr = GlobalRawType(*ptr);
+                            if ((grp_idx == groups[groups_pos]) && is_first_group_el) {
+                                groups_cache_.insert(groups[groups_pos]);
+                                val = curr;
+                            }
+                            else if (grp_idx != groups[groups_pos]) { // Updating when changing group
+                                if (is_first_group_el) { // Group not seen yet
+                                    groups_cache_.insert(groups[groups_pos]);
+                                    val = curr;
+                                }
+                                grp_idx = groups[groups_pos];
+                            }
+                        }
+                        ++groups_pos;
```

Comment:
Can this be moved to line 536, to make it clear it corresponds to the row we are in within this column?